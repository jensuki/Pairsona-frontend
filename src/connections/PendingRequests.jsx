import { useContext, useEffect, useState } from "react";
import { useNavigate, Link } from 'react-router-dom';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import Loading from '../components/Loading';
import noRequests from '../assets/placeholders/no_requests.png';
import '../styles/Connections.css';


const PendingRequests = () => {
    const { currUser, checkNotifications } = useContext(UserContext);
    const navigate = useNavigate();

    const [pendingRequests, setPendingRequests] = useState([]);
    const [loading, setLoading] = useState(true); // use local loading
    const [error, setError] = useState(null);

    // fetch pending requests whenever curr user changes
    useEffect(() => {
        if (!currUser) {
            setPendingRequests([]); // clear the list if the user logs out
            setLoading(false);
            return;
        }

        const fetchPendingRequests = async () => {
            try {
                setLoading(true);
                const res = await Api.getPendingRequests();
                setPendingRequests(res);
            } catch (err) {
                setError(err.message);
            } finally {
                setLoading(false);
            }
        };

        fetchPendingRequests();
    }, [currUser]);

    // handle accepting a connection request
    const handleAccept = async (connectionId) => {
        try {
            await Api.acceptConnectionRequest(connectionId);

            // remove conn request with given id from pending requests
            setPendingRequests((prev) =>
                prev.filter((req) => req.connectionId !== connectionId)
            );

            // redirect after accepting
            navigate('/connections');
        } catch (err) {
            setError(err.message);
        }
    };

    // handle declining a connection request
    const handleDecline = async (connectionId) => {
        try {
            await Api.declineConnectionRequest(connectionId);

            // remove declined request from the pending list
            setPendingRequests(pendingRequests.filter(req => req.connectionId !== connectionId));

            // refresh notifications immediately
            checkNotifications();
        } catch (err) {
            setError(err.message);
        }
    };

    if (error) return <p className="error">{error}</p>
    if (loading) return <Loading />

    return (
        <div className="PendingRequests">
            <h2>Connection Requests</h2>
            {pendingRequests.length === 0 ? (
                <img src={noRequests} alt="No Requests" />
            ) : (
                <ul className="pending-requests-list">
                    {pendingRequests.map((req) => (
                        <li key={req.connectionId} className="pending-request-item">
                            <div className="request-info">
                                <Link to={`/profile/${req.username}`} className="requester-name">
                                    <strong>{req.firstName} {req.lastName}</strong>
                                </Link>
                                <span> wants to connect with you</span>
                            </div>
                            {/* accept or decline buttons */}
                            <div className="request-actions">
                                <button className="accept-btn" onClick={() => handleAccept(req.connectionId)}>
                                    Accept
                                </button>
                                <button className="decline-btn" onClick={() => handleDecline(req.connectionId)}>
                                    Decline
                                </button>
                            </div>
                        </li>
                    ))}
                </ul>
            )
            }
        </div >
    );
};

export default PendingRequests;
