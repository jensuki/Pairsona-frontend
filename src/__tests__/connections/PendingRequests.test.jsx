import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import PendingRequests from '../../connections/PendingRequests';
import { expect, describe, test, vi, beforeEach } from 'vitest';
import Api from '../../helpers/api';
import { demoConnectionRequests, demoUser } from '../../test_common/mockData';

// mock api calls
vi.mock('../../helpers/api', () => ({
    default: {
        getPendingRequests: vi.fn(),
        acceptConnectionRequest: vi.fn(),
        declineConnectionRequest: vi.fn()
    }
}));

// helper
const renderPendingRequests = () => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={demoUser}>
                <PendingRequests />
            </UserProvider>
        </MemoryRouter>
    );
}

describe('PendingRequests component', () => {
    beforeEach(() => {
        vi.clearAllMocks();

        // common mock setup for pending requests
        Api.getPendingRequests.mockResolvedValue(demoConnectionRequests);
    })

    test('renders without crashing', async () => {
        Api.getPendingRequests.mockResolvedValue([]);

        renderPendingRequests();

        await waitFor(() => {
            expect(screen.getByText(/Connection Requests/i)).toBeInTheDocument();
        })
    })

    test('matches snapshot', async () => {
        const { asFragment } = renderPendingRequests();

        await waitFor(() => {
            expect(screen.getByText(/Connection Requests/i)).toBeInTheDocument();
        })
        // check if the mock is called
        expect(Api.getPendingRequests).toHaveBeenCalled();

        // verify snapshot
        expect(asFragment()).toMatchSnapshot();
    });

    test('displays no requests placeholder when there are no pending requests', async () => {
        Api.getPendingRequests.mockResolvedValue([]);

        renderPendingRequests();
        // expect alt from img to render
        await waitFor(() => {
            expect(screen.getByAltText(/No Requests/i)).toBeInTheDocument();
        });
    });

    test('displays pending requests', async () => {
        renderPendingRequests();

        // expect requesters first and last name to be rendered
        await waitFor(() => {
            demoConnectionRequests.forEach((req) => {
                // use regex match 'User One' exactly
                const fullName = new RegExp(`${req.firstName}\\s+${req.lastName}`, 'i');
                expect(screen.getByText(fullName)).toBeInTheDocument()
            })
        })
    });

    test('handles accepting a connection request', async () => {
        // mock api for accepting
        Api.acceptConnectionRequest.mockResolvedValue();

        renderPendingRequests();

        await waitFor(() => {
            expect(screen.getByText(/Connection Requests/i)).toBeInTheDocument();
        });

        // click the accept button
        const acceptBtn = screen.getByText(/Accept/i);
        fireEvent.click(acceptBtn);

        // verify mock api call was made with correct connection id
        expect(Api.acceptConnectionRequest)
            .toHaveBeenCalledWith(demoConnectionRequests[0].connectionId);

        // proceed to verify the pending requesters name is no longer in list
        await waitFor(() => {
            expect(screen.queryByText(demoConnectionRequests[0].firstName)).not.toBeInTheDocument();
        })
    })

    test('handles declining a connection request', async () => {
        renderPendingRequests();

        await waitFor(() => {
            expect(screen.getByText(/Connection Requests/i)).toBeInTheDocument();
        })

        // click decline button
        const declineBtn = screen.getByText(/Decline/i);
        fireEvent.click(declineBtn);

        // verify api was called with correct connection id
        expect(Api.declineConnectionRequest)
            .toHaveBeenCalledWith(demoConnectionRequests[0].connectionId);

        // proceed to verify the requesters name is no longer in list
        await waitFor(() => {
            expect(screen.queryByText(demoConnectionRequests[0].firstName)).not.toBeInTheDocument();
        })
    })

    test('displays error message when API call fails', async () => {
        // mock api with error msg
        Api.getPendingRequests.mockRejectedValue(new Error('Failed to fetch requests'));

        renderPendingRequests();

        // expect error
        await waitFor(() => {
            expect(screen.getByText(/Failed to fetch requests/i)).toBeInTheDocument();
        });
    });
})