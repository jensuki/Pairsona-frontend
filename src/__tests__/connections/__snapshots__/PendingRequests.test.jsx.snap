// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PendingRequests component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PendingRequests"
  >
    <h2>
      Connection Requests
    </h2>
    <ul
      class="pending-requests-list"
    >
      <li
        class="pending-request-item"
      >
        <div
          class="request-info"
        >
          <a
            class="requester-name"
            data-discover="true"
            href="/profile/user1"
          >
            <strong>
              User One
            </strong>
          </a>
          <span>
             wants to connect with you
          </span>
        </div>
        <div
          class="request-actions"
        >
          <button
            class="accept-btn"
          >
            Accept
          </button>
          <button
            class="decline-btn"
          >
            Decline
          </button>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
